File: Nov2020/25Nov/NextGreaterNodeInLinkedList.java
/*
Problem Name: 1019. Next Greater Node In Linked List
Problem Link: https://leetcode.com/problems/next-greater-node-in-linked-list/
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        ListNode temp = head;
        int n = 0;
        while(temp != null){
            temp = temp.next;
            n++;
        }
        int ans[] = new int[n], ele[] = new int[n];
        Stack<Integer> st = new Stack<Integer>();
        temp = head;
        n = 0;
        while(temp != null){
            if(st.isEmpty()){
                st.push(n);
                ele[n] = temp.val;
            }
            else{
                while(!st.isEmpty() && (ele[st.peek()] < temp.val)){
                    ans[st.pop()] = temp.val;
                }
                st.push(n);
                ele[n] = temp.val;
            }
            n++;
            temp = temp.next;
        }
        return ans;
    }
}

File: Nov2020/25Nov/NextGreaterNodeInLinkedList.cpp
/*
Problem name: 1019. Next Greater Node In Linked List
Problem link: https://leetcode.com/problems/next-greater-node-in-linked-list/
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        ListNode* cur = head;
        int len=0,i=0;
        while(cur != NULL){
            cur=cur->next;
            len++;
        }
        
        vector<int> ans(len,0);
        stack<pair<int,int>> s;
        
        cur=head;
        while(cur != NULL)
        {
            while(!s.empty() && cur->val > s.top().first)
            {
                ans[s.top().second] = cur->val;
                s.pop();
            }
            s.push(make_pair(cur->val, i));

            cur = cur->next;
            i++;
        }
        
        return ans;
    }
};
